### ISSUE TITLE
Getting Heartbeat Failed

### ISSUE THREAD

**Issue by vineetgoel** [aio-libs/aiokafka#335]
With one of my consumers which is subscribed to quite a few topics, I am hitting a point where I am getting the following errors repeatedly (seems like in a loop):

```
[2018-02-27 23:19:01,112: WARNING]: Heartbeat failed: local member_id was not recognized; resetting and re-joining group
[2018-02-27 23:19:01,117: ERROR]: Heartbeat session expired - marking coordinator dead
[2018-02-27 23:19:01,119: WARNING]: Marking the coordinator dead (node 107)for group edgy-v3.
[2018-02-27 23:19:01,338: ERROR]: ['Heartbeat failed: local member_id was not recognized; resetting and re-joining group']
[2018-02-27 23:19:01,344: ERROR]: ['Marking the coordinator dead (node 107)for group edgy-v3.']
```

Any idea why this might be happening and what might be causing this? The consumer is unable to recover. Also, looks like the particular broker is fine.

**Comment by bt-wil** [comment]
I've ran into the same issue, which make it so the consumers are unable to scale since it breaks all the consumer. 

The topic has 128 partitions, I'm trying to run ~12 consumers, but had a really hard time getting 4+ running at the same time.

when rebalancing starts: 
`Heartbeat failed for group <group-id> because it is rebalancing`

30 seconds later:
`Heartbeat failed: local member_id was not recognized; resetting and re-joining group"`

another 27 seconds later,
```
Heartbeat session expired - marking coordinator dead
Marking the coordinator dead (node 1008)for group
Discovered coordinator 1008 for group event-processor
Heartbeat failed: local member_id was not recognized; resetting and re-joining group
```

```
[2018-03-08_03:28:23.45757] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed for group event-processor because it is rebalancing
[2018-03-08_03:28:23.45786] INFO     aiokafka.consumer.group_coordinator: Revoking previously assigned partitions frozenset({...}) for group event-processor
[2018-03-08_03:28:26.45674] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed for group event-processor because it is rebalancing
[2018-03-08_03:28:29.45665] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed for group event-processor because it is rebalancing
[2018-03-08_03:28:32.45493] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed for group event-processor because it is rebalancing
[2018-03-08_03:28:35.45440] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed for group event-processor because it is rebalancing
[2018-03-08_03:28:38.45742] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed for group event-processor because it is rebalancing
[2018-03-08_03:28:41.45347] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed for group event-processor because it is rebalancing
[2018-03-08_03:28:44.45311] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed for group event-processor because it is rebalancing
[2018-03-08_03:28:47.45317] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed for group event-processor because it is rebalancing
[2018-03-08_03:28:50.45339] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed for group event-processor because it is rebalancing
[2018-03-08_03:28:53.45357] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed: local member_id was not recognized; resetting and re-joining group
[2018-03-08_03:28:53.55539] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed: local member_id was not recognized; resetting and re-joining group
[2018-03-08_03:28:53.65690] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed: local member_id was not recognized; resetting and re-joining group
[2018-03-08_03:28:53.76811] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed: local member_id was not recognized; resetting and re-joining group
[2018-03-08_03:28:53.86949] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed: local member_id was not recognized; resetting and re-joining group
...
...
[2018-03-08_03:29:20.25513] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed: local member_id was not recognized; resetting and re-joining group
[2018-03-08_03:29:20.35685] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed: local member_id was not recognized; resetting and re-joining group
[2018-03-08_03:29:20.45869] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed: local member_id was not recognized; resetting and re-joining group
[2018-03-08_03:29:20.45876] ERROR    aiokafka.consumer.group_coordinator: Heartbeat session expired - marking coordinator dead
[2018-03-08_03:29:20.45883] WARNING  aiokafka.consumer.group_coordinator: Marking the coordinator dead (node 1008)for group event-processor.
[2018-03-08_03:29:20.56109] INFO     aiokafka.consumer.group_coordinator: Discovered coordinator 1008 for group event-processor
[2018-03-08_03:29:20.56230] WARNING  aiokafka.consumer.group_coordinator: Heartbeat failed: local member_id was not recognized; resetting and re-joining group
...
...
```

What I also noticed is that for the process that has the heartbeat issue, it wasn't able to revoke the partitions due to a lock in my ConsumerRebalanceListener.  I think you can replicate this behavior by just "sleeping" inside rebalance listener.


Here is the simplified version of the code.
```
import asyncio
import logging
import aiokafka

logging.basicConfig(
    format='[%(asctime)-15s] [%(levelname)s]: %(message)s',
    level=logging.INFO, datefmt='%Y-%m-%d %H:%M:%S')

log = logging.getLogger()


class MyRebalancer(aiokafka.ConsumerRebalanceListener):
    def __init__(self, lock):
        self.lock = lock

    async def on_partitions_revoked(self, revoked):
        log.info('revoking, waiting on lock')
        # wait until a single batch processing is done
        async with self.lock:
            pass
        log.info('revoked')

    async def on_partitions_assigned(self, assigned):
        pass


class Server:
    def __init__(self, loop):
        self.lock = asyncio.Lock(loop=loop)
        self.consumer = aiokafka.AIOKafkaConsumer(
            loop=loop,
            bootstrap_servers=...,
            group_id='my-group-id',
            auto_offset_reset='latest',
            enable_auto_commit=False,
        )
        self.consumer.subscribe(
            topics=['topic'],
            listener=MyRebalancer(self.lock)
        )

    async def run(self):
        await self.consumer.start()
        while True:
            async with self.lock:
                await self.process()

    async def process(self):
        result = await self.consumer.getmany(timeout_ms=100, max_records=100)
        log.info('processing...')
        await asyncio.sleep(2)
        log.info('done processing')
        await self.consumer.commit()


loop = asyncio.get_event_loop()
server = Server(loop)
loop.run_until_complete(server.run())
```

For some reason, when the repartition happens during the the `await self.consumer.getmany`, it waits for the lock to become free during the `on_partitions_revoked`, but it never gets hold of the lock.  Is that a bug?

After fiddling around with it, I change the code to exclude the `consumer.getmany` from the lock block, and it seems to be working.  (I'm still going to do some test around it)
```
async def run(self):
    await self.consumer.start()
    while True:
        await self.process()

async def process(self):
    result = await self.consumer.getmany(timeout_ms=100, max_records=100)
    async with self.lock:
        log.info('processing...')
        await asyncio.sleep(2)
        log.info('done processing')
        await self.consumer.commit()
```

Is there a better way of handling the repartition assignment to make sure that a batch is processed (fetched/processed and committed) before repartition happens?

**Comment by tvoinarovskyi** [comment]
Ok, this is an interesting use case and is by all means valid. If you do a lock in the callback you will get the deadlock... For now, I don't have any ideas on how to get around this. This use case cannot happen in kafka-python nor Java client because those are synchronous and `getmany` would actually execute the callback, meaning no deadlock can happen.

**Comment by bt-wil** [comment]
I've switched over to exclude the `consumer.getmany` from the lock, and so far has been working.  I've been testing it by starting/stopping consumer(s) continuously and haven't encountered the heartbeat issue.  Does this mean that `consumer.getmany` is causing the deadlock?

As far as I understand what the flow should look like:
1. acquire the lock, and enter the `process()`
1. `await consumer.getmany()` yields the event loop
1. rebalancing happens, and consumer rebalance listener is called.
1. cannot `await for lock` in the listener, so it yields back control
1. consumer returns records
1. process records
1. exits lock
1. notifies the rebalance listener that the lock is free, which proceeds to rebalance.

As for deadlock in kafka-python, I thought it _would_ happen since it is synchronous. If the critical region is locked, and the consumer rebalance listener is called within the critical region, then wouldn't deadlock happen because the lock is not free, hence just waiting forever?

**Comment by tvoinarovskyi** [comment]
@bt-wil `consumer.getmany()` is a coroutine and will do IO to get messages from Kafka broker. Currently `aiokafka` sets it to wait for rebalance to finish, as we don't want to fetch messages on partitions, that may be assigned to someone else after rebalance.
`on_partitions_revoked` is created to allow user to finish processing of messages, that were **_already_** fetched, ie. to process and call commit. Trying to get new messages after rebalance has started will only slow down the process for all other consumers.
As for `kafka-python`, I ment that as it's single threaded you are not able to `resume` anything outside the callback. You have to do the logic inside the callback anyway, as other code will deadlock obviously.

We can't really warn the user about a potential deadlock, as the library does not know if `getmany` was called from callback or from main loop. Ie. I don't see any sane way to avoid the deadlock problem without changing the API =(. Ideas?

**Comment by Artimi** [comment]
@tvoinarovskyi I also have this issue. Going through the code I don't understand two things:
* Why does `aiokafka` send the `HeartbeatRequest` with `member_id == JoinGroupRequest.UNKNOWN_MEMBER_ID` when it is clear that it fails? 
* Shouldn't `GroupCoordinator._coordinator_routine` run in some loop? Because as I went through the code I saw that heartbeat is looping with `UNKNOWN_MEMBER_ID` and callingb`GroupCoordinator.reset_generation()` (which also calls `GroupCoordinator.request_rejoin()`), but until one of three conditions in `GroupCoordinator._coordinator_routine` (closing, coordinator dead, unsubscribe) is triggered nothing happens and therefore there will be no rejoin.

For me the problem with `GroupCoordinator` happens when a consumer unsubscribes from a topic. Then this loop happens until I subscribe a new topic.

**Comment by tvoinarovskyi** [comment]
@Artimi So for the questions:
* `aiokafka` sets `JoinGroupRequest.UNKNOWN_MEMBER_ID` cause we just don't have anything better to set. The idea is, that `_coordinator_routine` will rejoin right after we do it anyway.
* There is a 4th condition there:
```python
            # In case of manual assignment this future will be always set and
            # we don't want a heavy loop here.
            # NOTE: metadata changes are for partition count and pattern
            # subscription, which is irrelevant in case of user assignment.
            if auto_assigned:
                futures.append(self._rejoin_needed_fut)
```
We will rejoin right after reset_generation(), so it should not matter what is set in heartbeat task. 

> For me the problem with GroupCoordinator happens when a consumer unsubscribes from a topic. Then this loop happens until I subscribe a new topic.

You mean, heartbeat task should, probably, stop spamming logs in case of unsubscribe. Can we set that as a separate issue, this one is about a deadlock if we wait on some condition in ConsumerRebalanceListener.

**Comment by tvoinarovskyi** [comment]
This issue turned into several issues related one to another. Let me conclude some points and close this issue in favour of ones that splitted out of here. 

* @bt-wil pointed out a problem, that if a consumer's listener will do a lock acquire and processing code will also have the same lock, they will deadlock. This is expected and no matter how much I think, there is no good way to prevent such a situation. The related documentation was updated with a notice (https://github.com/aio-libs/aiokafka/commit/2656dbfae4df725ff4b68c4260c227cfcffa25a7#diff-589cbf3d563cf8d76673c3348a492da7R345).
* Confusion on error messages like ``Heartbeat failed for group <group-id> because it is rebalancing`` being WARNING level. Splitted out in a separate issue #461
* @Artimi was concerned that Consumer did not stop heartbear after unsubscribe called. That is covered by issue #372

**Comment by OpakAlex** [comment]
I changed group id, and it fixed issue. So maybe for somebody this will be useful.

**Comment by tfrokt** [comment]
I am running into this issue with `aiokafka==0.7.2` From the warning message it is not clear to me if this is actually an error or if I can just ignore it (if so, how?). 

`Heartbeat failed for group <group> because it is rebalancing`

It is also not clear to me if it's rebalancing and hence failing the heartbeat - or if it failed the heartbeat and hence it's rebalancing. In first case I would assume, I just don't care, in latter case there should be a solution to fix failing heartbeat.

Have read through various threads and SO posts and it's not clear to me what to do with this message. Would be great if someone with more experience could shine some light.

**Comment by MicaelJarniac** [comment]
Does it make sense to do it like the following?
```py
from contextlib import suppress

class MyRebalancer(aiokafka.ConsumerRebalanceListener):
    def __init__(self, lock):
        self.lock = lock

    async def on_partitions_revoked(self, revoked):
        await self.lock.acquire()

    async def on_partitions_assigned(self, assigned):
        with suppress(RuntimeError):
            self.lock.release()
```
I'm struggling to find a good example of what to do.
